lib/AST/RequirementMachine/RewriteSystem.cpp

```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
 #undef ASSERT_RULE
 }
 
+std::vector<RequirementError>
+RewriteSystem::getRedundantRequirements() {
+  // The resulting set of redundant requirements.
+  std::vector<RequirementError> redundantRequirements;
+
+  // Map redundant rule IDs to their rewrite path for easy access
+  // in the `isRedundantRule` lambda.
+  llvm::SmallDenseMap<unsigned, RewritePath> redundantRules;
```

Comment:
I doubt it actually matters for efficiency but it's a bit unfortunate to copy the RewritePath here since it's backed by an std::vector. Since the property of the rewrite path that you use is whether it contains an 'implied rule', can you precompute that in the map value instead?